ccm-connector/src/main/java/com/sequenceiq/cloudbreak/ccmimpl/endpoint/MinaSshdServiceEndpointFinder.java

```diff
@@ -0,0 +1,144 @@
+        map.put(KnownServiceIdentifier.GATEWAY, MinaSshdProto.EndpointService.Value.GATEWAY);
+        map.put(KnownServiceIdentifier.KNOX, MinaSshdProto.EndpointService.Value.KNOX);
+        ENDPOINT_SERVICE_VALUE_BY_KNOWN_SERVICE_IDENTIFIER = Maps.immutableEnumMap(map);
+    }
+
+    /**
+     * Those statuses that indicate a transient problem.
+     */
+    private static final Set<Status.Code> RETRYABLE_STATUS_CODES =
+            Sets.immutableEnumSet(Status.Code.UNAVAILABLE, Status.Code.NOT_FOUND);
+
+    private final ManagedChannel channel;
+
+    private final MinaSshdGrpc.MinaSshdBlockingStub blockingStub;
+
+    /**
+     * Creates a minasshd service endpoint Finder that will use the specified gRPC endpoint.
+     *
+     * @param lookupHost the hostname or ip for the minasshd gRPC endpoint
+     * @param lookupPort the listening port for the minasshd gRPC endpoint
+     */
+    public MinaSshdServiceEndpointFinder(String lookupHost, int lookupPort) {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(lookupHost), "Lookup host is required.");
+        Preconditions.checkArgument(lookupPort >= MIN_PORT && lookupPort <= MAX_PORT, "Invalid lookup port, not in [1,65535].");
+
+        channel = ManagedChannelBuilder.forAddress(lookupHost, lookupPort)
+                // JSA todo revisit Not using TLS since implementation is local loopback interface for Altus
+                .build();
+        blockingStub = MinaSshdGrpc.newBlockingStub(channel);
+    }
+
+    /**
+     * Attempts Orderly Shutdown on Managed Channel.
+     */
+    @PreDestroy
+    protected void shutdown() {
+        channel.shutdown();
+    }
+
+    @Nonnull
+    @Override
+    public <T extends ServiceEndpoint> T getServiceEndpoint(@Nonnull ServiceEndpointRequest<T> serviceEndpointRequest)
+            throws ServiceEndpointLookupException {
+
+        String instanceId = serviceEndpointRequest.getTargetInstance().getTargetInstanceId();
+        if (instanceId.isEmpty()) {
+            throw new IllegalArgumentException("Target instance ID is required.");
+        }
+
+        Optional<MinaSshdProto.EndpointService.Value> knownServiceIdentifier =
+                serviceEndpointRequest.getServiceFamily().getKnownServiceIdentifier()
+                        .map(ENDPOINT_SERVICE_VALUE_BY_KNOWN_SERVICE_IDENTIFIER::get);
+        if (!knownServiceIdentifier.isPresent()) {
+            throw new IllegalArgumentException("Service identifier is required.");
+        }
+
+        MinaSshdProto.GetServiceEndpointRequest request = MinaSshdProto.GetServiceEndpointRequest.newBuilder()
+                .setEndpointService(knownServiceIdentifier.get())
+                .setInstanceId(instanceId).build();
+
+        try {
+            MinaSshdProto.GetServiceEndpointResponse response = blockingStub.getServiceEndpoint(request);
+            if (response == null) {
+                throw new ServiceEndpointLookupException("Got null response from minasshd gRPC call", false);
```

Comment:
Nit, maybe for later?: Wondering if the exception message can be more prescriptive. So there was a null response, what does that mean went wrong? Feel free to leave for now.

Reply:
Will think about it. I'm not sure what to do about it, though.